Last-Update: 2014-12-04
Forwarded: https://github.com/quantcast/qfs/pull/51
Author: Dmitry Smirnov <onlyjob@member.fsf.org>
Description: minor spelling corrections

--- a/src/cc/meta/ChunkServer.h
+++ b/src/cc/meta/ChunkServer.h
@@ -824,9 +824,9 @@
     /// Boolean that tracks whether this server is down
     bool mDown;
 
     /// Is there a heartbeat message for which we haven't
-    /// recieved a reply yet?  If yes, dont' send one more
+    /// received a reply yet?  If yes, don't send one more
     bool   mHeartbeatSent;
 
     /// did we skip the sending of a heartbeat message?
     bool mHeartbeatSkipped;
--- a/src/cc/meta/LayoutManager.cc
+++ b/src/cc/meta/LayoutManager.cc
@@ -2183,9 +2183,9 @@
     }
     if (r.clusterKey != mClusterKey) {
         r.statusMsg = "cluster key mismatch:"
             " expect: "   + mClusterKey +
-            " recieved: " + r.clusterKey;
+            " received: " + r.clusterKey;
         r.status = -EBADCLUSTERKEY;
         return false;
     }
     if (mChunkServerMd5sums.empty() || find(
@@ -2193,9 +2193,9 @@
                 mChunkServerMd5sums.end(),
                 r.md5sum) != mChunkServerMd5sums.end()) {
         return true;
     }
-    r.statusMsg = "MD5sum mismatch: recieved: " + r.md5sum;
+    r.statusMsg = "MD5sum mismatch: received: " + r.md5sum;
     r.status    = -EBADCLUSTERKEY;
     return false;
 }
 
--- a/src/cc/tools/qfs_main.cc
+++ b/src/cc/tools/qfs_main.cc
@@ -4565,9 +4565,9 @@
     "deleted after it's copied.\n",
 
     "get", "[-ignoreCrc] [-crc] <src> <localdst>",
     "Copy files that match the file pattern <src>\n\t\t"
-    "to the local name.  <src> is kept.  When copying mutiple,\n\t\t"
+    "to the local name.  <src> is kept.  When copying multiple,\n\t\t"
     "files, the destination must be a directory.\n\t\t"
     "-crc and -ignoreCrc options are ignored.\n",
 /*
     "getmerge", "<src> <localdst>",
@@ -4637,9 +4637,9 @@
     "\n\t"
     "MODE\tMode is same as mode used for chmod shell command.\n\t\t"
     "Only letters recognized are 'rwxX'. E.g. a+r,g-w,+rwx,o=r\n"
     "\n\t"
-    "OCTALMODE\tMode specifed in 3 digits. Unlike shell command,\n\t\t"
+    "OCTALMODE\tMode specified in 3 digits. Unlike shell command,\n\t\t"
     "this requires all three digits.\n\t\t"
     "E.g. 754 is same as u=rwx,g=rx,o=r\n"
     "\n\t\t"
     "If none of 'augo' is specified, 'a' is assumed and unlike\n\t\t"
--- a/src/cc/common/kfserrno.cc
+++ b/src/cc/common/kfserrno.cc
@@ -165,9 +165,9 @@
         case EALREADY:         return kKFS_EALREADY;
         case EINPROGRESS:      return kKFS_EINPROGRESS;
         case EDQUOT:           return kKFS_EDQUOT;
         case ECANCELED:        return kKFS_ECANCELED;
-        // Ensure that the following do not clash with the sytem error values.
+        // Ensure that the following do not clash with the system error values.
         case EBADVERS:
         case ELEASEEXPIRED:
         case EBADCKSUM:
         case EDATAUNAVAIL:
--- a/src/cc/meta/ChunkServer.cc
+++ b/src/cc/meta/ChunkServer.cc
@@ -1211,9 +1211,9 @@
                     0 < mHelloOp->numNotStableAppendChunks ||
                     0 < mHelloOp->numNotStableChunks) &&
                 gLayoutManager.IsFileSystemIdRequired()) {
             KFS_LOG_STREAM_ERROR << GetPeerName() <<
-                " hello: invalid file sytem id" <<
+                " hello: invalid file system id" <<
             KFS_LOG_EOM;
             mHelloOp = 0;
             delete op;
             return -1;
